### PR TITLE
Heat laplace Operator on Digital Surfaces

### PR DESCRIPTION
**Pull request by cgurps**
# PR Description

The implementation in DGtal of the Laplace Beltrami operator for digital surfaces based on http://liris.cnrs.fr/david.coeurjolly/publications/caissard16preprint.html.

This PR contains the code for the operator, its associated documentation and some helpers to compute it.

The operator is embedded inside DEC as a LinearOperator between primal or dual 0-forms.
Steps to compute it requires 
* constructing a DEC with appropriates measures (one can do it by hands or use the new function createFromNScells in the DEC factory which takes the normal estimator cell functor and the grid step and automatically computes size of dual and primal n-forms).
* calling the new heatLaplace function of DEC. The function takes as argument the grid step, the time parameter for the convolution and the cut locus for the integration. Details are explained in the module Heat Kernel Laplace-Beltrami Operator in the DEC package from DGtal documentation.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [ ] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: src/DGtal/dec/DiscreteExteriorCalculus.h [hunk 1]
```diff
@@ -400,6 +401,17 @@ namespace DGtal
     LinearOperator<Self, 0, duality, 0, duality>
     laplace() const;
 
+    /**
+     * Convolutional Laplace operator from duality 0-forms to duality 0-forms. See \ref secHowToHeatLaplace.
+     * @param h the grid step
```
Comment (dcoeurjo) [inline src/DGtal/dec/DiscreteExteriorCalculus.h]: please add a cite and an entry in the bibtex pointing to your DGCI paper.

File: src/DGtal/dec/DiscreteExteriorCalculus.h [hunk 2]
```diff
@@ -400,6 +401,17 @@ namespace DGtal
...
+     * @param t the time parameter for the convolution
+     * @param cut_locus the cut locus for for the integration. Generaly setted to \f$ \log( - \log( t ) + 1 ) + 2 \f$
```
Comment (dcoeurjo) [inline src/DGtal/dec/DiscreteExteriorCalculus.h]: for for -> for
setted->set

the "cut locus" is not very clear... please rephrase.

File: src/DGtal/dec/DiscreteExteriorCalculus.h (hunk 1 above)
Comment (dcoeurjo) [reply src/DGtal/dec/DiscreteExteriorCalculus.h]: from dual 0-forms to dual 0-froms

File: src/DGtal/dec/DiscreteExteriorCalculusFactory.h [hunk 3]
```diff
@@ -103,6 +103,23 @@ class DiscreteExteriorCalculusFactory
     DiscreteExteriorCalculus<dimEmbedded, TNSCellConstIterator::value_type::Point::dimension, TLinearAlgebraBackend, TInteger>
     createFromNSCells(const TNSCellConstIterator& begin, const TNSCellConstIterator& end, const bool add_border = true);
 
+    /**
+     * Create a DEC structure from a range of signed n-cells, where n is the embedded dimension. A functor from cell to its measure is given.
+		 * Signed n-cells may live in an ambient Khamlisky space with dimension greater than n.
```
Comment (dcoeurjo) [inline src/DGtal/dec/DiscreteExteriorCalculusFactory.h]: tab

File: src/DGtal/dec/DiscreteExteriorCalculusFactory.h [hunk 4]
```diff
@@ -143,6 +160,24 @@ class DiscreteExteriorCalculusFactory
...
+     * @param cell starting cell.
+     * @param cells_accum cells accumulator in which lower incident cells get inserted and counted.
+     * @param local_accum local cells accumulator for measure embedding
+     * @param cell_to_measure a map from cell to its measure
+     * @param measure the current measure to be embedded
+     */
+    template <typename KSpace, typename CellsAccum, typename MeasureAccum>
+    static
+    void
+    accumulateAllLowerIncidentCells(const KSpace& kspace, const typename CellsAccum::key_type& cell, CellsAccum& cells_accum, CellsAccum& local_accum, MeasureAccum& cell_to_measure, const double measure);
```
Comment (dcoeurjo) [inline src/DGtal/dec/DiscreteExteriorCalculusFactory.h]: these methods weren't already in the helpers ?

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+//----------------------------------------
+/*!
+@page moduleHeatLaplaceOperator Heat Kernel Laplace-Beltrami Operator
+@writers Caissard Thomas
+
+[TOC]
+
+\section sectHeatIntro Continuous Laplace-Beltrami operator
+
+Let \f$M\f$ be a smooth manifold. The linear operator \f$\Delta : C^2(\partial M) \rightarrow C^2(\partial M)\f$ (where \f$C^2(\partial M)\f$ is the set of twice differentiable function acting on \f$\partial M\f$)
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: please reformat to 80col

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+For example it is, in \f$ \mathbb{R}^d \f$, the operator is the sum of the second derivatives:
+\f[
+    \Delta_{\mathbb{R}^d} u(x) = \sum_i \frac{\partial^2}{\partial x_i^2}u(x).
+\f]
+This operator can be expressed in local coordinates on a manifold given the metric tensor \f$g_{ij}\f$:
+\f[
+    \Delta u = \frac{1}{\sqrt{g}} \partial_i \left( \sqrt{|g|}g^{ij}\partial_j u \right),
+\f]
+with the einstein notation implied (meaning that repeated indexes are summed over), \f$ \sqrt{|g|}\f$ is the absolute value of the determinant of \f$ g_{ij} \f$
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: Einstein

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+\f]
+Using a finite difference on \f$t\f$, and the basic property that the integral
+of the heat kernel must be one:
+\f[
+    \Delta g(x, t) = \lim\limits_{t \rightarrow 0} \frac{1}{t} \int_{y\in \partial M } \tilde{p}(t, x, y) (u(y) - u(x))dy.
+\f]
+
+\section secDiscre The Discretization on Digital Surfaces
+
+We adapt the formulation of Belkin et al. on digital surfaces.
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: add refs and cites... (cf bibtex file for the entries).

Comment (dcoeurjo) [reply src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: ok you have added couple of entries -:) please add Belkin and your DGCI papers.

**Comment by dcoeurjo** [comment]
thanks for the PR. Please add a uni test file (with catch). Just checking properties of the operator (cols sum to 1...)

File: src/DGtal/dec/DiscreteExteriorCalculusFactory.h (hunk 4 above)
Comment (cgurps) [reply src/DGtal/dec/DiscreteExteriorCalculusFactory.h]: The automatic construction of cell measures using a normal estimator was not provided by the factory (although you can do it by hand).

**Comment by dcoeurjo** [comment]
Could you please keep me updated with this PR (Release pending) ? Thanks

File: src/DGtal/dec/DiscreteExteriorCalculus.h (hunk 1 above)
Comment (cgurps) [reply src/DGtal/dec/DiscreteExteriorCalculus.h]: the operator is valid from any duality

**Comment by cgurps** [comment]
I'm currently making tests for the operator

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+Let \f$M\f$ be a smooth manifold. The linear operator \f$\Delta : C^2(\partial
+M) \rightarrow C^2(\partial M)\f$ (where \f$C^2(\partial M)\f$ is the set of
+twice differentiable function acting on \f$\partial M\f$) defined by 
+\f[ \Delta u = \nabla^2 u = \nabla \cdot \nabla u, \f] is called the Laplace-Beltrami
+operator. 
+@note The sign of the operator is arbitrary, and one may found in the
+litterature the alternative definition \f$-\nabla \cdot \nabla u\f$ for
+\f$\Delta u\f$.
+
+For example it is, in \f$ \mathbb{R}^d \f$, the operator is the sum of the
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: remove first "it is"

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+second derivatives: \f[ \Delta_{\mathbb{R}^d} u(x) = \sum_i
+\frac{\partial^2}{\partial x_i^2}u(x). \f] This operator can be expressed in
+local coordinates on a manifold given the metric tensor \f$g_{ij}\f$: \f[ \Delta
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: local coordinates system

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+tangent plane \f$T_p M\f$ at a point \f$p\f$.
+
+The operator can also be written in a coordinate-free manner using the exterior
+calculus theory. Given a function \f$w\f$ (ie. a 0-form) we have \f[ \Delta w =
+\star d \star d w, \f] where \f$\star\f$ is the hodge operator and \f$ d \f$ the
+exterior derivative.
+
+\section subBelkin The Laplace-Beltrami operator from the Heat Equation
+
+Let \f$g : \partial M \times (0, T) \rightarrow \mathbb{R}\f$ be a
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: Add refs in the global.bib and doxygen `@cite`

Comment (dcoeurjo) [reply src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: Ok I see below

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+which is the initial temperature distribution.
+
+\image html heat_diffusion_manifold.png "The heat diffusion on a manifold. The heat kernel is represented in green, the geodesic distance in blue."
+
+An exact solution is: \f[ \int_{y\in \partial M} p(t, x, y) u(y) dy, \f] where
+\f$p \in C^\infty(\mathbb{R}^+ \times \partial M \times \partial M)\f$ is the
+heat kernel \cite rosenberg.
+
+There is a wide range of studies on the behavior of \f$p\f$ when \f$t\f$ tends
+to zero (called small time asymptotics of diffusion process). Early work
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: Early works include

Comment (dcoeurjo) [reply src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: Earlier?

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+:= \frac{1}{t_h(4 \pi t_h)^{\frac{d}{2}}} \sum_{\mathbf{r}\in \mathbb{F}^d }
+e^{- \frac{||\dot{\mathbf{r}}
+        - \dot{\mathbf{s}}||^2}{4t_h}}  [f(\dot{{\mathbf{r}}})
+    - f(\dot{{\mathbf{s}}})]\mu(\mathbf{r}), \f] where \f$\mathbb{F}^d\f$ is the
+      set of elements dimension \f$d\f$ (for example surfels in a 2D surface
+      embedded in 3D), \f$\dot{\mathbf{r}}\f$ (resp. \f$\dot{\mathbf{s}}\f$) is
+      the centroid of the surfel \f$\mathbf{r}\f$ (resp. \f$\mathbf{s}\f$), and
+      \f$\mu(\mathbf{r})\f$ is equal to the dot product between an estimated
+      normal and the trivial normal orthogonal to the surfel \f$\mathbf{s}\f$.
+
```
Comment (dcoeurjo) [inline src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: you should add a note saying that $alpha=1/3$ as in II would do the job and cite the JMIV paper (arxiv?HAL?) with the formal details and proofs

File: tests/dec/testHeatLaplace.cpp [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+  return test_operator( laplace_primal.myContainer ) && test_operator( laplace_dual.myContainer );
+}
+
+TEST_CASE( "Operator Test" )
+{
+  SECTION( "Test" );
+  {
+    REQUIRE( laplace_sphere( 1.0 ) );
+    REQUIRE( laplace_sphere( 0.5 ) );
+    REQUIRE( laplace_sphere( 0.2 ) );
```
Comment (dcoeurjo) [inline tests/dec/testHeatLaplace.cpp]: ok.. not great but ok..

**Comment by dcoeurjo** [comment]
Please update the Changelog as well

File: src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox (hunk 12 above)
Comment (cgurps) [reply src/DGtal/dec/doc/moduleHeatLaplaceOperator.dox]: I uploaded the paper to HAL as a preprint and waiting for their approval.

**Comment by cgurps** [comment]
Appart from the reference to the JMIV paper, this PR is ready !

**Comment by dcoeurjo** [comment]
Ok let's wait for HAL, thanks

**Comment by dcoeurjo** [comment]
I'll merge this PR. We'll fix the HAL ref later;
Thanks

**Pull request merged**